### PR TITLE
Dump viper on cache hit (version 2)

### LINKED ISSUE #843: Cached programs do not dump their Viper encoding
When using the Prusti Server with `PRUSTI_DUMP_VIPER_PROGRAMS=true`, cached programs do not write their Viper encoding to disk. We might want that to happen anyway.

### PR DESCRIPTION
**Pull request by fpoli**
Fixes https://github.com/viperproject/prusti-dev/issues/843.

An alternative to https://github.com/viperproject/prusti-dev/pull/845.

### CODE REVIEW

File: prusti-server/src/process_verification.rs [hunk 1]
```diff
@@ -10,67 +10,70 @@ use prusti_common::{config, report::log::report, vir::ToViper, Stopwatch};
...
+
+        // Print the hash and skip verification. Used for testing.
+        if config::print_hash() {
+            println!(
+                "Received verification request for: {}",
+                request.program.get_name()
+            );
+            println!("Hash of the request is: {}", hash);
+            // Some tests need the dump to report a diff of the Viper programs.
+            let _ = build_or_dump_viper_program();
```
Comment (JonasAlaif) [inline prusti-server/src/process_verification.rs]: There is no need to run this if `!config::dump_viper_program()`

File: prusti-server/src/process_verification.rs [hunk 2]
```diff
@@ -10,67 +10,70 @@ use prusti_common::{config, report::log::report, vir::ToViper, Stopwatch};
...
+pub fn process_verification_request<'v, 't: 'v>(
     verification_context: &'v VerificationContext<'t>,
     request: VerificationRequest,
     cache: impl Cache,
 ) -> viper::VerificationResult {
-    if !config::enable_cache() || config::print_hash() {
-        process_verification_request(verification_context, request)
-    } else {
+    let ast_utils = verification_context.new_ast_utils();
+    ast_utils.with_local_frame(16, || {
```
Comment (JonasAlaif) [inline prusti-server/src/process_verification.rs]: I'm not sure how expensive `with_local_frame` is? If it's cheap then this is fine, but if not then it's problematic for the common case where `!config::dump_viper_program()` but we get a cache hit

Comment (fpoli) [reply prusti-server/src/process_verification.rs]: I've not benchmarked it, but it should be cheaper than doing a method call in the JVM.

Comment (fpoli) [reply prusti-server/src/process_verification.rs]: (avoided)

**Review by JonasAlaif — APPROVED** [review]

**Pull request merged**